src/overlay/OverlayManagerImpl.cpp

```diff
@@ -517,6 +519,46 @@ OverlayManagerImpl::connectTo(std::vector<PeerBareAddress> const& peers,
     return count;
 }
 
+void
+OverlayManagerImpl::maybeDropRandomPeer()
+{
+    // If we haven't heard from the network for a while, try randomly
+    // disconnecting a peer in hopes of picking a better one. (preferred peers
+    // aren't affected as we always want to stay connected)
+    auto now = mApp.getClock().now();
+    if (!mApp.getHerder().isTracking())
+    {
+        if (mLastOutOfSyncReconnect)
+        {
+            // We've been out of sync, check if it's time to reconnect
+            if (now - *mLastOutOfSyncReconnect > OUT_OF_SYNC_RECONNECT_DELAY)
+            {
+                auto allPeers = getOutboundAuthenticatedPeers();
+                std::vector<std::pair<NodeID, Peer::pointer>> nonPreferredPeers;
+                std::copy_if(std::begin(allPeers), std::end(allPeers),
+                             std::back_inserter(nonPreferredPeers),
+                             [&](auto const& peer) {
+                                 return !mApp.getOverlayManager().isPreferred(
+                                     peer.second.get());
+                             });
+                if (!nonPreferredPeers.empty())
+                {
+                    auto peerToDrop = rand_element(nonPreferredPeers);
+                    peerToDrop.second->sendErrorAndDrop(
+                        ERR_LOAD, "random disconnect due to out of sync",
+                        Peer::DropMode::IGNORE_WRITE_QUEUE);
+                }
+            }
+            else
+            {
+                // Still waiting for the timeout
+                return;
+            }
+        }
+    }
+    mLastOutOfSyncReconnect = std::make_optional<VirtualClock::time_point>(now);
```

Comment:
Should this be moved to when we actually drop a peer?

Reply:
I think it's not the only place where we should set the time: when we go first go out of sync, we want to timeout before this logic kicks in (in case there's a transient issue unrelated to disconnected graph).

Reply:
I am not sure I understand:
I thought the intent of `mLastOutOfSyncReconnect` was to throttle how often we drop peers.

Why would we want to wait the very first time? When that code triggers we already have connected to a lot of peers (because we don't have capacity) so I don't think this would be a transient issue.

If we want to avoid doing certain things right when we lose sync, we need to reset `mLastOutOfSyncReconnect` when we lose sync. That said the node was already waiting `CONSENSUS_STUCK_TIMEOUT_SECONDS` (35 seconds) before going out of sync, so I don't think the extra delay is needed (ie: when we lose sync, it's possible that we're only connected to bad peers).

Reply:
I don't think the extra delay is controversial, it's a conservative change to prevent unnecessary peer churn because "not tracking" does not always equal "bad connection graph".

Reply:
Which comment are you responding to? I was asking which problem you're trying to address as it looks like you're trying to do something related to when a node transitions to out of sync but it's potentially incomplete

Reply:
I'm talking about waiting for timeout before dropping peer for the first time after going out of sync. `mLastOutOfSyncReconnect` is reset either way in that function, whether we're tracking or not tracking. I'm not sure what you're saying above about the logic being incomplete.